src/main/java/beam/physsim/cch/RoutingFrameworkTravelTimeCalculator.scala

```diff
@@ -0,0 +1,251 @@
+package beam.physsim.cch
+
+import java.util
+import java.util.concurrent.Executors
+
+import beam.agentsim.events.PathTraversalEvent
+import beam.sim.BeamServices
+import com.google.common.util.concurrent.ThreadFactoryBuilder
+import com.typesafe.scalalogging.LazyLogging
+import com.vividsolutions.jts.geom.{Coordinate, Envelope}
+import com.vividsolutions.jts.index.kdtree.{KdNode, KdTree}
+import org.apache.commons.lang.time.StopWatch
+import org.matsim.api.core.v01.network.Link
+import org.matsim.core.controler.events.IterationEndsEvent
+
+import scala.collection.JavaConverters._
+import scala.collection.convert.ImplicitConversionsToScala._
+import scala.collection.mutable
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+class RoutingFrameworkTravelTimeCalculator(
+  private val beamServices: BeamServices,
+  private val osmInfoHolder: OsmInfoHolder,
+  private val routingFrameworkWrapper: RoutingFrameworkWrapper,
+) extends LazyLogging {
+
+  private implicit val executionContext: ExecutionContext = ExecutionContext.fromExecutorService(
+    Executors.newFixedThreadPool(
+      Math.max(Runtime.getRuntime.availableProcessors() / 4, 1),
+      new ThreadFactoryBuilder().setDaemon(true).setNameFormat("routing-framework-worker-%d").build()
+    )
+  )
+
+  private val odsFactor: Int =
+    Math.max(
+      (1.0 / beamServices.beamConfig.beam.agentsim.agentSampleSizeAsFractionOfPopulation *
+      beamServices.beamConfig.beam.physsim.cch.congestionFactor).toInt,
+      1
+    )
+
+  private val graph: RoutingFrameworkGraph = routingFrameworkWrapper.generateGraph()
+
+  private val coord2VertexId: Map[Coordinate, Long] =
+    graph.vertices.map(x => x.coordinate -> x.id).toMap
+
+  private val kdTree = new KdTree()
+  graph.vertices.foreach { x =>
+    kdTree.insert(x.coordinate, x.id)
+  }
+
+  def generateLink2TravelTimes(
+    pathTraversalEvents: util.Collection[PathTraversalEvent],
+    iterationNumber: Int,
+    links: util.Collection[_ <: Link],
+    maxHour: Int
+  ): java.util.Map[String, Array[Double]] = {
+
+    val travelTimeCalculationStopWatch = new StopWatch()
+
+    val id2Link = links.toStream.map(x => x.getId.toString.toInt -> x).toMap
+
+    val hour2Events = generateHour2Events(pathTraversalEvents)
+
+    val odsCreationFutures = hour2Events
+      .map {
+        case (hour, events) =>
+          Future {
+            val stopWatch: StopWatch = new StopWatch
+            stopWatch.start()
+
+            var odNumber = 0
+
+            val ods = generateOdsFromTravelInfos(events, id2Link)
+
+            val odStream = ods
+              .flatMap { od =>
+                odNumber = odNumber + 1
+                Stream.range(0, odsFactor, 1).map(_ => od)
+              }
```

Comment:
(moving my outdated comment here) You can use `Stream.fill` if you want to be lazy - now that you are manipulating a var you can't replace the whole statement, but you could do it in the end part

Reply:
fixed